csrc/scheduler/mma_utils.cpp

```diff
@@ -1731,6 +1729,96 @@ std::unordered_map<ValGroup, MatmulDomain> MatmulPattern::getDimRoles(
   return dim_roles;
 }
 
+std::vector<ValGroup> canonicalDimOrdering(
+    const mma_utils::TensorRolesMap& tensor_roles,
+    const mma_utils::DimRolesMap& dim_roles,
+    const ValGraph& exact_graph) {
```

Comment:
In the exact graph, broadcasting IDs are not mapped to concrete IDs. So how will the result of this function look like, will it be like `(batch, broadcasting M, non-broadcasting M, broadcasting N, non-broadcasting N, K)`? Could you write you reply to this question as a comment in the code?

Reply:
Good question. I have so far only used it to determine which concrete dims are innermost in which case we can ignore broadcasts. However, when we actually use this to do the reordering it will be important so I will change it to use the permissive map, and add a comment explaining this.